File: app/src/main/java/ir/arapp/arappmain/view/utils/CustomCardView.java
package ir.arapp.arappmain.view.utils;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.util.AttributeSet;


import com.google.android.material.card.MaterialCardView;

import ir.arapp.arappmain.R;

public class CustomCardView extends MaterialCardView {

    private boolean canChange = true;
    public float getCornerRadiusByPercent() {
        return cornerRadiusByPercent;
    }

    //value must be between 0f..1f
    public void setCornerRadiusByPercent(float cornerRadiusByPercent) {
        this.cornerRadiusByPercent = cornerRadiusByPercent;
        postInvalidate();
    }

    void changeCornerSize(){
        canChange = true;
        postInvalidate();
    }

    float cornerRadiusByPercent = -1f;

//    public float getCornerRadiusBySize() {
//        return cornerRadiusBySize;
//    }
//
//    public void setCornerRadiusBySize(float cornerRadiusBySize) {
//        this.cornerRadiusBySize = cornerRadiusBySize;
//        postInvalidate();
//    }
//
//    float cornerRadiusBySize = 0f;


    public CustomCardView(Context context) {
        super(context);
    }

    public CustomCardView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    public CustomCardView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs);
    }

    private void init(Context context, AttributeSet attrs) {
        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.CustomCardView);
        cornerRadiusByPercent = typedArray.getFloat(R.styleable.CustomCardView_cardCornerRadiusByPercent, -1f);
        typedArray.recycle();
        changeCornerSize();
    }


    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        changeCornerSize();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if (canvas != null) {
            if (canChange){
                int width = this.getWidth();
                int height = this.getHeight();
                if (cornerRadiusByPercent >= 0 && cornerRadiusByPercent <= 1)
                    this.setRadius((float) (Math.min(width , height) * cornerRadiusByPercent));
                canChange = false;
            }

        }
    }



}

File: app/src/main/java/ir/arapp/arappmain/view/utils/CustomCardView.kt
package ir.arapp.arappmain.view.utils

import android.content.Context
import android.graphics.Canvas
import android.util.AttributeSet
import com.google.android.material.card.MaterialCardView
import ir.arapp.arappmain.R

class CustomCardView : MaterialCardView {
    private var canChange = true

    //value must be between 0f..1f


    fun changeCornerSize() {
        canChange = true
        postInvalidate()
    }

    var cornerRadiusByPercent = -1f
    set(value) {
        field = value
        postInvalidate()
    }

    //    public float getCornerRadiusBySize() {
    //        return cornerRadiusBySize;
    //    }
    //
    //    public void setCornerRadiusBySize(float cornerRadiusBySize) {
    //        this.cornerRadiusBySize = cornerRadiusBySize;
    //        postInvalidate();
    //    }
    //
    //    float cornerRadiusBySize = 0f;
    constructor(context: Context?) : super(context) {}
    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(context, attrs)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        init(context, attrs)
    }

    private fun init(context: Context, attrs: AttributeSet) {
        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.CustomCardView)
        cornerRadiusByPercent =
            typedArray.getFloat(R.styleable.CustomCardView_cardCornerRadiusByPercent, -1f)
        typedArray.recycle()
        changeCornerSize()
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        changeCornerSize()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (canvas != null) {
            if (canChange) {
                val width = this.width
                val height = this.height
                if (cornerRadiusByPercent >= 0 && cornerRadiusByPercent <= 1) this.radius =
                    (Math.min(width, height) * cornerRadiusByPercent)
                canChange = false
            }
        }
    }
}